NMS-8894: Add transient attribute to Measurements API (#5947)
OpenNMS/opennms: docs/modules/development/pages/rest/measurements.adoc

= Measurements API

> You can use the Measurements API to retrieve collected values stored in RRD or JRB files, and in Newts.

Reviewer: ```suggestion
You can use the Measurements API to retrieve collected values stored in RRD or JRB files, in Newts, and in other supported time-series databases.
```

Actual step may differ.
| 300000

| maxrows
| When using the API to render a graph, this should be set to the graph's pixel width.
| 0

| aggregation
| Consolidation function used (`AVERAGE`, `MIN`, `MAX`).
The available options depend on the RRA definition.
| AVERAGE

| fallback-attribute
| Secondary attribute that is queried if the primary attribute does not exist.
| _Blank_
|===

=== Step sizes

The `step` parameter behaves differently depending on the time series strategy that your {page-component-title} instance uses.

When using persistence strategies based on RRD, the available step sizes are limited to those defined by the RRA when the file was created.
The effective step size used is one that covers the requested period and is closest to the requested step size.
For maximum accuracy, we recommend that you use a step size of `1`.

When using Newts, you can set the step size arbitrarily; this is because the aggregation is performed at request time.
To prevent large requests, {page-component-title} limits the step size to a maximum of 5 minutes (the default collection rate).
You can decrease this value by modifying `org.opennms.newts.query.minimum_step`.

=== cURL example

> The following example demonstrates how the GET function and its parameters can be used with cURL.

Reviewer 2: ```suggestion
The following example demonstrates how you can use the GET function and its parameters with cURL.
```

| Timestamp.
If > `0`, the timestamp is relative to the UNIX epoch (January 1st 1970 00:00:00 AM). +
If < 0, the timestamp is relative to the `end` option (for example, default value is 4 hours ago).
| -14400000

| end
| Timestamp.
If \<= `0`, the effective value is the current timestamp.
| 0

| step
| Requested time interval between rows.
Actual step may differ.
| 300000

| maxrows
| When using the API to render a graph, this should be set to the graph's pixel width.
| 0

| aggregation
| Consolidation function used (`AVERAGE`, `MIN`, `MAX`).
The available options depend on the RRA definition.
| AVERAGE

| fallback-attribute
| Secondary attribute that is queried if the primary attribute does not exist.
| _Blank_
|===

=== Step sizes

The `step` parameter behaves differently depending on the time series strategy that your {page-component-title} instance uses.

When using persistence strategies based on RRD, the available step sizes are limited to those defined by the RRA when the file was created.
The effective step size used is one that covers the requested period and is closest to the requested step size.
For maximum accuracy, we recommend that you use a step size of `1`.

> When using Newts, you can set the step size arbitrarily; this is because the aggregation is performed at request time.

Author: I would assume this means that you can set the value to an arbitrary number. @mmahacek, could you confirm?

> When using Newts, you can set the step size arbitrarily; this is because the aggregation is performed at request time.

Reviewer: I think Emily's read of this is the correct one.

> When using Newts, you can set the step size arbitrarily; this is because the aggregation is performed at request time.

Author: ```suggestion
When using Newts, you can set the step size to an arbitrary number; this is because the aggregation is performed at request time.
```

            "attribute": "ifHCInOctets",
            "label": "ifHCInOctets",
            "resourceId": "nodeSource[Servers:1424038123222].interfaceSnmp[eth0-04013f75f101]",
            "transient": "false" <1>
        },
        {
            "aggregation": "AVERAGE",
            "attribute": "ifHCOutOctets",
            "label": "ifHCOutOctets",
            "resourceId": "nodeSource[Servers:1424038123222].interfaceSnmp[eth0-04013f75f101]",
            "transient": "true" <2>
        }
    ],
    "expression": [
        {
            "label": "ifHCOutOctetsNeg",
            "value": "-1.0 * ifHCOutOctets",
            "transient": "false" <1>
        }
    ]
}
----
<1> Indicates that the attribute is not included in the query output.

> <2> Indicates that the attribute is needed for query input.

Author: Thanks! I'll update these.

> <2> Indicates that the attribute is needed for query input.

Reviewer 3: To be more accurate, the transient settings determine if the attribute is included in the response or not.  It technically doesn't mean if the attribute is needed or not. Any attribute referenced in an expression query must be requested individually.  When setting transient to `true`, it is a way to hide the source attribute in the response so you only get the calculated expression result back.

| `java.lang.Math.PI`

| __diff_time
| Time span between the start and end of samples.

| __step
| Difference in time between subsequent values.

| __i

> | Index into the samples array that the present calculation is referencing.

Reviewer 2: ```suggestion
| Index into the samples array that the present calculation references.
```

| __AttributeName (where `AttributeName` is the searched-for attribute)
| This returns the complete `double[]` array of samples for `AttributeName`.
|===

{page-component-title} predefines a number of functions that you can use in expressions.
They are referenced by `namespace:function`, and they all return a Java double value:

[cols="2,3,1"]
|===
| Function  | Description   | Example

| jexl:evaluate("_formula"):

> | Passes a string to the JEXL engine to be evaluated as if it were entered as a normal expression. +

Reviewer 2: ```suggestion
| Passes a string to the JEXL engine to evaluate as if it were entered as a normal expression. +
```
